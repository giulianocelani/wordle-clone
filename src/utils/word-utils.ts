import words from '../5-letters.json';

const MAX_WORD_LENGTH = 5;

function getRandomWord() {
	return words[Math.floor(Math.random() * words.length)];
}

export { getRandomWord, MAX_WORD_LENGTH };
