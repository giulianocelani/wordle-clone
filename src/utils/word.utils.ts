import words from '../5-letters.json';

const MAX_WORD_LENGTH = 5;

enum LetterState {
	MISS = 'MISS',
	MATCH = 'MATCH',
	PRESENT = 'PRESENT'
}

function getRandomWord() {
	return words[Math.floor(Math.random() * words.length)].toUpperCase();
}

function evaluateGuess(answer: string, guess: string): LetterState[] {
	if (answer.length !== guess.length && guess.length !== MAX_WORD_LENGTH) {
		return [];
	}

	const result: LetterState[] = Array(MAX_WORD_LENGTH).fill(LetterState.MISS);

	const lettersToCheck = answer.split('');
	const guessLetters = guess.split('');

	guessLetters.forEach((letter, i) => {
		if (answer[i] === letter) {
			result[i] = LetterState.MATCH;
			lettersToCheck.splice(i === lettersToCheck.length ? i - 1 : i, 1);
		}
	});

	guessLetters.forEach((letter, i) => {
		if (lettersToCheck.includes(letter) && result[i] !== LetterState.MATCH) {
			result[i] = LetterState.PRESENT;
			lettersToCheck.splice(lettersToCheck.indexOf(letter), 1);
		}
	});

	return result;
}

function isValidWord(word: string): boolean {
	return words.includes(word.toLowerCase());
}

export {
	evaluateGuess,
	getRandomWord,
	isValidWord,
	LetterState,
	MAX_WORD_LENGTH
};
