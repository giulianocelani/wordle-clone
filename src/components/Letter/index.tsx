import { mappings } from '../../constants';
import useStore from '../../store';
import { LetterState } from '../../utils';

type IProps = {
	letter: string;
	state?: LetterState;
};

const Letter = ({ letter, state }: IProps) => {
	const { keysPressed } = useStore();

	const getWrapperStyle = () => {
		const styles: string[] = [];
		if (state) {
			styles.push(mappings.StateToColor[state]);
		} else if (keysPressed[letter] === LetterState.MISS) {
			styles.push('bg-gray-100 border-gray-100');
		}
		return styles.join(' ');
	};

	return (
		<div
			className={`border rounded-md flex items-center justify-center ${getWrapperStyle()}`}
		>
			<h2
				className={`font-semibold text-xl md:text-2xl lg:text-3xl ${
					state || keysPressed[letter] === LetterState.MISS ? 'text-white' : ''
				}`}
			>
				{letter}
			</h2>
		</div>
	);
};

Letter.defaultProps = {
	state: undefined
};

export default Letter;
