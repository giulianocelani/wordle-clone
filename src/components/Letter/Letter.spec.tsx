import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { LetterState } from '../../utils';

import Letter from '.';

describe('Letter', () => {
	let wrapper = mount(<Letter letter='A' />);

	it('default renders correctly', () => {
		expect(toJson(wrapper, { mode: 'deep' })).toMatchSnapshot();

		expect(wrapper.exists('.bg-green-500')).toBeFalsy();
		expect(wrapper.exists('.border-green-500')).toBeFalsy();
		expect(wrapper.exists('.bg-yellow-500')).toBeFalsy();
		expect(wrapper.exists('.border-yellow-500')).toBeFalsy();
		expect(wrapper.exists('.bg-gray-500')).toBeFalsy();
		expect(wrapper.exists('.border-gray-500')).toBeFalsy();

		expect(wrapper.find('h2').exists('.text-white')).toBeFalsy();
	});

	it('renders letter correctly', () => {
		expect(wrapper.find('h2').text()).toEqual('A');

		wrapper.setProps({ letter: 'B' });
		expect(wrapper.find('h2').text()).toEqual('B');
	});

	it('MATCH renders correctly', () => {
		wrapper = mount(<Letter letter='A' state={LetterState.MATCH} />);
		expect(toJson(wrapper, { mode: 'deep' })).toMatchSnapshot();

		expect(wrapper.exists('.bg-green-500')).toBeTruthy();
		expect(wrapper.exists('.border-green-500')).toBeTruthy();

		expect(wrapper.exists('.bg-yellow-500')).toBeFalsy();
		expect(wrapper.exists('.border-yellow-500')).toBeFalsy();
		expect(wrapper.exists('.bg-gray-500')).toBeFalsy();
		expect(wrapper.exists('.border-gray-500')).toBeFalsy();

		expect(wrapper.find('h2').exists('.text-white')).toBeTruthy();
	});

	it('MISSING renders correctly', () => {
		wrapper = mount(<Letter letter='A' state={LetterState.MISS} />);
		expect(toJson(wrapper, { mode: 'deep' })).toMatchSnapshot();

		expect(wrapper.exists('.bg-gray-500')).toBeTruthy();
		expect(wrapper.exists('.border-gray-500')).toBeTruthy();

		expect(wrapper.exists('.bg-green-500')).toBeFalsy();
		expect(wrapper.exists('.border-green-500')).toBeFalsy();
		expect(wrapper.exists('.bg-yellow-500')).toBeFalsy();
		expect(wrapper.exists('.border-yellow-500')).toBeFalsy();

		expect(wrapper.find('h2').exists('.text-white')).toBeTruthy();
	});

	it('EXISTS renders correctly', () => {
		wrapper = mount(<Letter letter='A' state={LetterState.PRESENT} />);
		expect(toJson(wrapper, { mode: 'deep' })).toMatchSnapshot();

		expect(wrapper.exists('.bg-yellow-500')).toBeTruthy();
		expect(wrapper.exists('.border-yellow-500')).toBeTruthy();

		expect(wrapper.exists('.bg-green-500')).toBeFalsy();
		expect(wrapper.exists('.border-green-500')).toBeFalsy();
		expect(wrapper.exists('.bg-gray-500')).toBeFalsy();
		expect(wrapper.exists('.border-gray-500')).toBeFalsy();

		expect(wrapper.find('h2').exists('.text-white')).toBeTruthy();
	});

	it('renders correctly with no letter', () => {
		wrapper = mount(<Letter letter='' />);
		expect(wrapper.find('h2').text()).not.toEqual('A');
	});
});
