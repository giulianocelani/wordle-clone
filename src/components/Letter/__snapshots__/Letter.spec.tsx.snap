// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Letter EXISTS renders correctly 1`] = `
<div
  className="border rounded-md flex items-center justify-center bg-yellow-500 border-yellow-500"
>
  <h2
    className="font-semibold text-xl md:text-2xl lg:text-3xl text-white"
  >
    A
  </h2>
</div>
`;

exports[`Letter MATCH renders correctly 1`] = `
<div
  className="border rounded-md flex items-center justify-center bg-green-500 border-green-500"
>
  <h2
    className="font-semibold text-xl md:text-2xl lg:text-3xl text-white"
  >
    A
  </h2>
</div>
`;

exports[`Letter MISSING renders correctly 1`] = `
<div
  className="border rounded-md flex items-center justify-center bg-gray-500 border-gray-500"
>
  <h2
    className="font-semibold text-xl md:text-2xl lg:text-3xl text-white"
  >
    A
  </h2>
</div>
`;

exports[`Letter default renders correctly 1`] = `
<div
  className="border rounded-md flex items-center justify-center "
>
  <h2
    className="font-semibold text-xl md:text-2xl lg:text-3xl "
  >
    A
  </h2>
</div>
`;
